sys/hashes/sha3.c

```diff
@@ -0,0 +1,463 @@
+ */
+void FIPS202_SHAKE128(const unsigned char *input, unsigned int inputByteLen, unsigned char *output, int outputByteLen)
+{
+    Keccak(1344, 256, input, inputByteLen, 0x1F, output, outputByteLen);
+}
+
+/**
+ *  Function to compute SHAKE256 on the input message with any output length.
+ */
+void FIPS202_SHAKE256(const unsigned char *input, unsigned int inputByteLen, unsigned char *output, int outputByteLen)
+{
+    Keccak(1088, 512, input, inputByteLen, 0x1F, output, outputByteLen);
+}
+
+/**
+ *  Function to compute SHA3-224 on the input message. The output length is fixed to 28 bytes.
+ */
+void FIPS202_SHA3_224(const unsigned char *input, unsigned int inputByteLen, unsigned char *output)
+{
+    Keccak(1152, 448, input, inputByteLen, 0x06, output, 28);
+}
+
+
+/**
+ *  Function to compute SHA3-256 on the input message. The output length is fixed to 32 bytes.
+ */
+void FIPS202_SHA3_256(const unsigned char *input, unsigned int inputByteLen, unsigned char *output)
+{
+    Keccak(1088, 512, input, inputByteLen, 0x06, output, SHA3_256_DIGEST_LENGTH);
+}
+
+void sha3_256(void *digest, const void *data, size_t len)
+{
+    FIPS202_SHA3_256(data, len, digest);
+}
+
+void sha3_256_init(keccak_state_t *ctx)
+{
+    Keccak_init(ctx, 1088, 512, 0x06);
+}
+
+void sha3_update(keccak_state_t *ctx, const void *data, size_t len)
+{
+    Keccak_update(ctx, data, len);
+}
+
+void sha3_256_final(keccak_state_t *ctx, void *digest)
+{
+    Keccak_final(ctx, digest, SHA3_256_DIGEST_LENGTH);
+}
+
+
+/**
+ *  Function to compute SHA3-384 on the input message. The output length is fixed to 48 bytes.
+ */
+void FIPS202_SHA3_384(const unsigned char *input, unsigned int inputByteLen, unsigned char *output)
+{
+    Keccak(832, 768, input, inputByteLen, 0x06, output, 48);
+}
+
+void sha3_384(void *digest, const void *data, size_t len)
+{
+    FIPS202_SHA3_384(data, len, digest);
+}
+
+void sha3_384_init(keccak_state_t *ctx)
+{
+    Keccak_init(ctx, 832, 768, 0x06);
+}
+
+void sha3_384_final(keccak_state_t *ctx, void *digest)
+{
+    Keccak_final(ctx, digest, SHA3_384_DIGEST_LENGTH);
+}
+
+/**
+ *  Function to compute SHA3-512 on the input message. The output length is fixed to 64 bytes.
+ */
+void FIPS202_SHA3_512(const unsigned char *input, unsigned int inputByteLen, unsigned char *output)
+{
+    Keccak(576, 1024, input, inputByteLen, 0x06, output, 64);
+}
+
+void sha3_512(void *digest, const void *data, size_t len)
+{
+    FIPS202_SHA3_512(data, len, digest);
+}
+
+void sha3_512_init(keccak_state_t *ctx)
+{
+    Keccak_init(ctx, 576, 1024, 0x06);
+}
+
+void sha3_512_final(keccak_state_t *ctx, void *digest)
+{
+    Keccak_final(ctx, digest, SHA3_512_DIGEST_LENGTH);
+}
+
+/*
+   ================================================================
+   Technicalities
+   ================================================================
+ */
+
+typedef unsigned char UINT8;
+typedef unsigned long long int UINT64;
```

Comment:
This might not be platform independent. Rather use `uint8_t` and `uint64_t` from `<stdint.h>`.